keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,823 @@
+  option (alpha_message) = true;
+
+  message Entry {
+    // This field is REQUIRED
+    GroupSnapshot group_snapshot = 1;
+  }
+
+  repeated Entry entries = 1;
+
+  // This token allows you to get the next page of entries for
+  // `ListGroupSnapshots` request. If the number of entries is larger than
+  // `max_entries`, use the `next_token` as a value for the
+  // `starting_token` field in the next `ListGroupSnapshots` request. This
+  // field is OPTIONAL.
+  // An empty string is equal to an unspecified field value.
+  string next_token = 2;
+}
+```
+
+## Proposal for Volume Placement
+
+### API Changes
+
+In order to support Volume Placement, An `AllowedTopologies` field will be added to the VolumeGroupClass API:
+
+```
+type VolumeGroupClass struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+
+        // Driver is the driver expected to handle this VolumeGroupClass.
+        // This value may not be empty.
+        Driver string
+
+        // Parameters holds parameters for driver.
+        // These values are opaque to the system and are passed directly
+        // to the driver.
+        // +optional
+        Parameters map[string]string
+
+        // This field specifies whether group snapshot is supported.
+        // The default is false.
+        // +optional
+        GroupSnapshot *bool
+
+        // Restrict the topologies where a group of volumes can be located.
+        // Each driver defines its own supported topology specifications.
+        // An empty TopologySelectorTerm list means there is no topology restriction.
+        // This field is passed on to the drivers to handle placement of a group of
+        // volumes on storage pools.
+        // +optional
+        AllowedTopologies []api.TopologySelectorTerm
+}
+```
+
+### Example Yaml Files for Volume Placement
+
+A VolumeGroupClass that supports placement:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupClass
+metadata:
+  name: placementGroupClass1
+spec:
+  parameters:
+     …...
+  allowedTopologies: [failure-domain.example.com/placement: storagePool1]
+```
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroup
+metadata:
+  Name: placemenGroup1
+spec:
+  volumeGroupClassName: placementGroupClass1
+```
+
+A PVC that belongs to both the volume group with groupSnapshot support and placement.
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: pvc1
+  annotations:
+spec:
+  accessModes:
+  - ReadWriteOnce
+  dataSource: null
+  resources:
+        requests:
+        storage: 1Gi
+  storageClassName: storageClass1
+  volumeMode: Filesystem
+  volumeGroupNames: [volumeGroup1, placementGroup1]
+```
+
+If both placement group and volume group with groupSnapshot support are defined, it is possible for the same volume to join both groups. For example, a volume group with groupSnapshot support may include volume members from two placement groups as they belong to the same application.
```

Comment:
Is it possible that one volume can join multiple volume groups with groupSnapshot and multiple placement groups?

Reply:
The current design allows that.

Reply:
I feel a little bit confuse if one volume could join multiple volume groups with conflict placement policy. Or the placement polices would never conflict with each other?

Reply:
That would definitely make the controller more complicated.  The controller will have to make sure they are compatible, otherwise the group should not be created successfully.